### PR TITLE
Implemented a Siamese Network Example

### PR DESCRIPTION
**Pull request by sudomaze**
I have followed `mnist/main.py` setup and modified it a bit to include the definition of `Siamese Network` and custom configuration of `MNIST` set but for the `Object Matching` task.

This PR was created per https://github.com/pytorch/examples/issues/645

### CODE REVIEW

**Comment by msaroufim** [comment]
Can you please setup a test here as well? https://github.com/pytorch/examples/blob/main/run_python_examples.sh

File: siamese_network/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Basic Siamese Network Example
+
+```bash
+pip install -r requirements.txt
+python main.py
+# CUDA_VISIBLE_DEVICES=2 python main.py  # to specify GPU id to ex. 2
+```
```
Comment (msaroufim) [inline siamese_network/README.md:7]: Can you also reference the paper you're using a baseline for your implementation?

File: siamese_network/main.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        output = self.resnet(x)
+        output = output.view(output.size()[0], -1)
+        return output
+
+    def forward(self, input1, input2):
+        # get two images' features
+        output1 = self.forward_once(input1)
+        output2 = self.forward_once(input2)
+
+        # concatnate both images' features
```
Comment (msaroufim) [inline siamese_network/main.py:60]: theres's a few simple typos around, a quick spell checker would help

File: siamese_network/main.py [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    test_loss /= len(test_loader.dataset)
+
+    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
+        test_loss, correct, len(test_loader.dataset),
+        100. * correct / len(test_loader.dataset)))
+
+
+def main():
+    # Training settings
+    parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
```
Comment (msaroufim) [inline siamese_network/main.py:158]: Update to Siamese network

File: siamese_network/main.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        return output
+
+class APP_MATCHER(Dataset):
+    def __init__(self, root, train, download=False):
+        super(APP_MATCHER, self).__init__()
+        # get MNIST dataset
+        self.dataset = datasets.MNIST(root, train=train, download=download)
+
+        # get targets (labels) and data (images)
+        self.targets = copy.deepcopy(self.dataset.targets)
```
Comment (msaroufim) [inline siamese_network/main.py:76]: what's the goal of the group sets section? Could you add some comments explaining why the deep copies are needed and what this is doing exactly?

File: siamese_network/main.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        for (images_1, images_2, targets) in test_loader:
+            images_1, images_2, targets = images_1.to(device), images_2.to(device), targets.to(device)
+            outputs = model(images_1, images_2).squeeze()
+            test_loss += criterion(outputs, targets).sum().item()  # sum up batch loss
+            pred = torch.where(outputs > 0.5, 1, 0)  # get the index of the max log-probability
+            correct += pred.eq(targets.view_as(pred)).sum().item()
+
+    test_loss /= len(test_loader.dataset)
+
+    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
```
Comment (msaroufim) [inline siamese_network/main.py:151]: Add comment around how much loss is expected to be if default settings are used

File: siamese_network/main.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    def group_sets(self):
+        np_arr = np.array(self.dataset.targets.clone())
+        self.grouped_indices = {}
+        for i in range(0,10):
+            self.grouped_indices[i] = np.where((np_arr==i))[0]
+    
+    def __len__(self):
+        return self.data.shape[0]
+    
+    def __getitem__(self, index):
```
Comment (msaroufim) [inline siamese_network/main.py:90]: add more comments in this section

**Review by msaroufim — CHANGES_REQUESTED** [review]

File: siamese_network/README.md (hunk 1 above)
Comment (sudomaze) [reply siamese_network/README.md:7]: I have referenced `FaceNet`, the closest implementation to the example's implementation. This implementation varies from FaceNet as we use the `ResNet-18` model as our feature extractor. In addition, we aren't using `TripletLoss` as the MNIST dataset is simple, so `BCELoss` can do the trick.

File: siamese_network/main.py (hunk 2 above)
Comment (sudomaze) [reply siamese_network/main.py]: Thanks for pointing it out! I have fixed it! 😄

File: siamese_network/main.py (hunk 3 above)
Comment (sudomaze) [reply siamese_network/main.py]: Oh, I missed it. Thanks!

File: siamese_network/main.py (hunk 4 above)
Comment (sudomaze) [reply siamese_network/main.py]: I have removed unwanted copying of objects and included detailed comments to explain the process.

File: siamese_network/README.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# Siamese Network Example
```
Comment (msaroufim) [inline siamese_network/README.md:1]: OK last thing can you please quickly explain what a Siamese network is here, the high-level architecture and what users should expect to input and get as an output for this example. Not a tutorial.

Also, post some proof in a screenshot here on the PR that this thing works

**Review by msaroufim — APPROVED** [review]

File: siamese_network/README.md (hunk 7 above)
Comment (sudomaze) [reply siamese_network/README.md:1]: I have included an explanation in the class definition. I will update the read me to include the same definition.

Here is a screenshot to prove that the model works using the default setup.

Screenshot from 2022-05-10 21-44-58
